factom_keys/serverid.py

```diff
@@ -0,0 +1,293 @@
+CHECKSUM_LENGTH = 4
+BODY_LENGTH = 35
+TOTAL_LENGTH = 39
+
+
+def generate_id_key_set():
+    """
+    Generate a set of Factom Server Identity keypairs, sk1-sk4 and id1-id4.
+    :return: a list of tuples containing a random id key pair in string format: [(sk1..., id1...), (sk2..., id2...)]
+    """
+    pairs = []
+
+    for i in range(1, 5):
+        signer, verifier = ed25519.create_keypair()
+        k = ServerIdPrivateKey(signer.to_seed())
+        p = ServerIdPublicKey(verifier.to_bytes())
+        k.key_level = i
+        p.key_level = i
+        pairs.append((k.to_string(), p.to_string()))
+
+    return pairs
+
+
+def generate_key_pair(key_level: int):
+    """
+    Generate a Factom ServerId keypair at the specified level: 1-4 for (sk1, id1) - (sk4, id4).
+
+    :param key_level: the key_level as an integer, 1-4
+    :return: a tuple containing a random id key pair: (ServerIdPrivateKey, ServerIdIPublicKey)
+    """
+    signer, verifier = ed25519.create_keypair()
+    k = ServerIdPrivateKey(signer.to_seed())
+    k.key_level = key_level
+    p = ServerIdPublicKey(verifier.to_bytes())
+    p.key_level = key_level
+    return (k, p)
+
+
+def _to_base58_string(prefixed_key: bytes):
+    """
+    Convert prefixed_key bytes into sk1/id strings with a checksum.
+
+    :param prefixed_key: the sk1 private key or id public key prefixed with the appropriate bytes
+    :return: a sk1 private key string or id public key string
+    """
+    prefix = prefixed_key[:PREFIX_LENGTH]
+    if not (
+        prefix in ServerIdPublicKey.PREFIXES or prefix in ServerIdPrivateKey.PREFIXES
+    ):
+        raise InvalidPrefixError
+    temp_hash = sha256(prefixed_key[:BODY_LENGTH]).digest()
+    checksum = sha256(temp_hash).digest()[:CHECKSUM_LENGTH]
+    return base58.encode(prefixed_key + checksum)
+
+
+class BadKeyStringError(Exception):
+    pass
+
+
+class BadKeyBytesError(Exception):
+    pass
+
+
+class InvalidKeyLevelError(Exception):
+    pass
+
+
+class InvalidPrefixError(Exception):
+    pass
+
+
+class InvalidParamsError(Exception):
+    def __init__(self):
+        Exception.__init__(self, "Only provide one of seed bytes or key_string, not both")
+
+
+class ServerIdPrivateKey:
+
+    # Prefixes for sk1 - sk4.
+    PREFIXES = [b"\x4d\xb6\xc9", b"\x4d\xb6\xe7", b"\x4d\xb7\x05", b"\x4d\xb7\x23"]
+
+    def __init__(self, seed_bytes=None, key_string=None):
+        # Check that we only get one of the two params
+        if not ((seed_bytes and not key_string) or (
+            not seed_bytes and key_string
+        )):
+            raise InvalidParamsError
+        # Default to lowest key level.
+        self.key_level = 4
+
+        if key_string:
+            if not ServerIdPrivateKey.is_valid(key_string):
+                raise BadKeyStringError
+
+            try:
+                # Get level from third character
+                self.key_level = int(key_string[2])
+            except ValueError:
+                raise BadKeyStringError
```

Comment:
I think we can remove this safety check here since we already checked that its a valid key a few lines up, so that exception should be impossible to hit. Can just simplify this into: `self.key_level = int(key_string[2])